[Rewrite] Wiki base structure (#52)
QuiltMC/developer-wiki: wiki/blocks/first-block.md

---
title: Adding a Simple Block
index: 1
---
# Adding a Simple Block

Creating a block is quite similar to [creating an item](../items/first-item), but now leveraging both the block *and* item registries, as well as a more in-depth process for creating a model.

## Creating and Registering the Block

First we create the block and store it in a field:

```java

> public final Block EXAMPLE_BLOCK = new Block(new QuiltBlockSettings())

Reviewer: static, semicolon

```

Then we register it in the `onInitialize()` function:

```java
Registry.register(Registries.BLOCK, new Identifier(mod.metadata().id(), "example_block"), EXAMPLE_BLOCK);
```
## Adding a BlockItem

Having done that, we can place the block using the `setblock` command, but there isn't an item to place it. So we additionially register a `BlockItem` for the block and add it to the `BUILDING_BLOCKS` item group:

```java
Registry.register(Registries.ITEM, new Identifier(mod.metadata.id(), "example_block"), new BlockItem(EXAMPLE_BLOCK, new QuiltItemSettings()))

ItemGroupEvents.modifyEntriesEvent(ItemGroups.BUILDING_BLOCKS).register(entries -> {
	entries.addStack(EXAMPLE_BLOCK.asItem().getDefaultStack());
});
```

For more information on what this does, see the [Creating your first Item](../items/first-item#registering-the-item) Article.

## Adding a Model for the Block

First, we need to create the BlockStates file for your Block. This tells the game which BlockStates use which models. For an example on using a more complicated BlockState setup, see [Adding redstone functionality to your Block](TODO). In this case, we only have one BlockState and can use this simple json:

> ```json

Reviewer 2: ```suggestion
`assets/simple_block_mod/blockstates/example_block.json`:
```json
```
Moved path from after code block and added file name.
Also added `assets/simple_block_mod/` because I think it's important to make the path as clear as possible. Asset paths are a common issue, especially for new modders.

{
    "variants": {
        "": {
            "model": "simple_block:block/example_block"
        }
    }
}
```

Use this for your `example_block.json` in `blockstates` in your assets folder and replace `simple_block` and `example_block` with your mod id and block name.

Now we can put the model at the path specified by the identifier (in this case `models/block/example_block.json`).

```json
{
	"parent": "minecraft:block/cube_all",
	"textures": {
		"all": "simple_block:block/example_block"
	}
}
```

> 

Reviewer 2: ```suggestion
Replace `simple_block_mod` and `example_block` as before.

```

This will use the texture located at `textures/block/example_block` in your assets folder for all sides of the block.

To use this model for the item variant, too, declare the Block model as a parent in the item model which is located at `models/item/example_block.json` in your assets folder:
```json
{
	"parent": "simple_block:block/example_block"
}
```

> 

Reviewer 2: ```suggestion
Replace `simple_block_mod` and `example_block` as before.

```

## Adding a translation for the Block

Last but not least we need to add a translation for the block. This will automatically apply to the BlockItem, too.
```json
{
    "block.simple_block.example_block": "Example Block"
}
```

> 

Reviewer 2: ```suggestion
Replace `simple_block_mod` and `example_block` as before.

```

---
title: Adding your first Block
index: 1
---
# Adding your first Block

> Creating a Block is pretty similar to creating an Item, [read the corresponding article](../items/first-item) for more elaborate explanations.

Reviewer 3: ```suggestion
Creating a Block is pretty similar to creating an Item, read [the corresponding article](../items/first-item) for more elaborate explanations.
```
not a fan of this entire sentence honestly (I don't like saying "x is just like y!" without giving any reasons), I'd prefer something more like:
```
Creating a Block is quite similar to creating an [item](../items/first-item), but now leveraging both the block *and* item registries, as well a more in-depth process for creating a model.
```

## Creating and registering the Block

First we create the Block and store it in a field:

```java
public final Block EXAMPLE_BLOCK = new Block(new QuiltBlockSettings())
```

Then we register it in the `onInitialize()` function:

```java
Registry.register(Registries.BLOCK, new Identifier(mod.metadata().id(), "example_block"), EXAMPLE_BLOCK);
```

## Adding a BlockItem

> Having done that, we can place the block using the `setblock` command, but there isn't an item to place it. So we additionally register a `BlockItem` for the block and add it to the `BUILDING_BLOCKS` item group:

Reviewer 3: ```suggestion
Having done that, we can place the block using the `setblock` command but, opening the creative menu, we won't be able to find an item version of it. To fix this, we register a `BlockItem` for the block and add it to an item group, in our case `BUILDING_BLOCKS`:
```
cleaner language + avoiding the possibility of someone thinking that every block has to go in that group

```java
Registry.register(Registries.ITEM, new Identifier(mod.metadata().id(), "example_block"), new BlockItem(EXAMPLE_BLOCK, new QuiltItemSettings()))

ItemGroupEvents.modifyEntriesEvent(ItemGroups.BUILDING_BLOCKS).register(entries -> {
	entries.addItem(EXAMPLE_BLOCK.asItem());
});
```

For more information on what this does, see the [Creating your first Item](../items/first-item#registering-the-item) Article.

## Adding a Model for the Block

> First, we need to create the blockstates file for your Block. This tells the game which `BlockState`s use which models. For an example on using a more complicated `BlockState` setup, see [Adding Redstone functionality to your Block](redstone-interaction). In this case, we only have one `BlockState` and can use this simple JSON:

Reviewer 3: ```suggestion
First, we need to create the "blockstates" file for your Block. A `BlockState` is a different form that your block can take: for example, each stage of growth in a crop block is a different `BlockState`. This tells the game which `BlockState`s use which models. For an example on using a more complicated `BlockState` setup, see [Adding Redstone functionality to your Block](redstone-interaction). In this case, we only have one possible `BlockState` and can use this simple JSON:
```
clarifying what block states are

`assets/simple_block_mod/blockstates/example_block.json`:

```json
{
	"variants": {
		"": {
			"model": "simple_block_mod:block/example_block"
		}
	}
}
```

> Replace `simple_block_mod` and `example_block` with your mod id and block name. Be sure to replace them in both the file/folder names and the JSON file.

Reviewer 3: I'd toss this, we aren't telling the readers to change the names anywhere else, meaning it's possible someone sees this, renames the json nonsense, but leaves it registered as `example_block`

Reviewer 2: Rather than get rid of this, I think we should specify it everywhere it's needed.
Failing to update these in all the necessary places seems like a very common 'gotcha' for new modders.

Reviewer 3: ^ I agree with this actually. please add it to the block registration!

---

Now we can put the model at the path specified by the identifier (in this case `assets/simple_block_mod/models/block/example_block.json`).

```json
{
	"parent": "minecraft:block/cube_all",
	"textures": {
		"all": "simple_block_mod:block/example_block"
	}
}
```

Replace `simple_block_mod` and `example_block` as before.
This will use the texture located at `assets/simple_block_mod/textures/block/example_block.png` for all sides of the block.

---

> To use this model for the block item too, declare the Block model as a parent in the item model which is located at `assets/simple_block_mod/models/item/example_block.json`:

Reviewer 3: ```suggestion
Our block item needs a model too, but instead of simply using a texture as we did in the [item tutorial](../items/first-item), we'll declare the Block model as a parent in the item model (located at `assets/simple_block_mod/models/item/example_block.json`). This will render the block item identically to how the block would look in-game.
```
clarification on what this does and why

```json
{
	"parent": "simple_block_mod:block/example_block"
}
```

> Replace `simple_block_mod` and `example_block` as before.

Reviewer 3: previous comment applies

---
title: Adding a Simple Block
index: 1
---
# Adding a Simple Block

> Creating a block is quite similar to [creating an item](../items/first-item), but now leveraging both the block *and* item registries, as well a more in-depth process for creating a model.

Reviewer 2: ```suggestion
Creating a block is quite similar to [creating an item](../items/first-item), but now leveraging both the block *and* item registries, as well as a more in-depth process for creating a model.
```
